src/libAtomVM/externalterm.c

```diff
@@ -221,7 +223,56 @@ static void write_bytes(uint8_t *buf, avm_uint64_t val)
     }
 }
 
-static int serialize_term(uint8_t *buf, term t, GlobalContext *glb)
+static bool has_extended_utf8_encoding(atom_ref_t atom_ref, size_t atom_len)
+{
+    for (size_t i = 0; i < atom_len; ) {
+        size_t out_len = 0;
+        uint32_t c;
+        enum UnicodeTransformDecodeResult res = bitstring_utf8_decode(
+            (const uint8_t *) atom_ref,
```

Comment:
`atom_ref_t` is an opaque that points to an atom table node, I'm very surprised that this code is even working. Are we testing it?

Reply:
Nice catch.  It is, indeed well tested -- for example, see the universal quantifier test -- and I am surprised that not even a compiler warning was emitted (and that it passed).

In any event, I have fixed this to use the accessor for AtomStrings, which should be more correct.

Reply:
> Nice catch. It is, indeed well tested -- for example, see the universal quantifier test -- and I am surprised that not even a compiler warning was emitted (and that it passed).
> 
> In any event, I have fixed this to use the accessor for AtomStrings, which should be more correct.

what surprise me is that tests were passing, that might mean we are not having enough coverage.
So if you think we are covering all new code enough feel free to resolve this conversation, otherwise let's add some additional tests.

Reply:
Indeed, it turned out that this code was not executed, because the test never forced `{minor_version, 1}` for the atoms with UTF-8 chars.  I fixed that, and discovered a bug (also fixed) in the process, whereby atoms with leading non-unicode extended chars would be incorrectly encoded.  (The pointer over the AtomStr was not being incremented)